include/zephyr/sys/kref.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KREF_H_
+#define ZEPHYR_INCLUDE_SYS_KREF_H_
+
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup kref_apis Reference count APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when the count drops to 0. This can also be used on SMP systems.
+ */
+struct k_ref {
```

Comment:
Naming nitpick: you probably want to call this "sys_ref" and put it in lib/os and not kernel/.  The "k_" prefix and kernel are for (heh) code that interacts with the core kernel bits like threads, timing, memory management, etc...

This is actually just a lockless library layer built on top of the atomics API.

Reply:
I wasn't sure if we want this to be an implementation detail or not. For example, there is a variant of the libatomic implementation that implements them using spinlocks(which need kernel communication). You're probably right though.

Reply:
I renamed them